streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/CommonStreamsMetrics.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals.metrics;
+
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.Count;
+import org.apache.kafka.common.metrics.stats.Rate;
+import org.apache.kafka.common.metrics.stats.Total;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import static org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsConventions.threadLevelSensorName;
+import static org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsConventions.threadLevelTags;
+
+public final class CommonStreamsMetrics {
```

Comment:
I had another meta question while discussing offline, and I'll leave it here for discussion: since different meters belong to the different metrics levels, should we move these static functions to the corresponding `XXMetrics` instead? I'm thinking about that since we have some meters that exist in multiple levels, like commit rate. In the future if we move them all to this class then we need to get one function for each level as their tags naming conventions are different, and hence it is not really `common` streams metrics.

Reply:
There are two different concepts of belonging in play here.
1. grouping: a metric "belongs" to the Thread level if it's aggregated at that level
2. ownership: a metric "belongs" to the Thread level if StreamThread needs a reference to it.

Grouping and Ownership are orthogonal here.

There are several metrics that are aggregated at the thread level and are also owned by StreamThread, like the commitTimeSensor. Grouping is easy to determine; all you have to do is look at the metric name. To tell that commitTimeSensor is Owned by StreamThread, you can trace the references to the sensor. It's never passed outside of StreamThread, so it's owned by StreamThread.

Other metrics are still Grouped by thread but Owned by other classes, such as the taskCreatedSensor. All of StreamThread, AbstractTaskCreator, TaskCreator, and StandbyTaskCreator have references to this metric, so they all share the ownership. Now that I'm looking at it, StreamThread never uses the reference, so it could be owned only by AbstractTaskCreator and its children.

Finally, we get to the skippedRecordsSensor. This metric is still Grouped by thread, but Ownership is shared among 14 classes. Two of them are framework classes (GlobalStreamThread and StreamThread), and 12 of them are user-space Processor classes.

Since Ownership for skippedRecordsSensor is so dispersed, I decided to just make its ownership global, aka common. Anything in the entire Streams codebase can get a reference to the skippedRecordsSensor.

I wouldn't move all the metrics into the global ownership space just because I moved one, and I wouldn't give StreamThread a reference to a metric just because that metric happens to be aggregated by thread.

Am I making sense?